File: sprint5/By lessons/lesson10code1/lesson10code1.py
def heap_add(heap, key):
    index = len(heap) + 1
    heap.append(key)
    sift_up(heap, index)

def sift_up(heap, index):
    if index == 1:
        return

    parentIndex = index // 2
    if heap[parentIndex - 1] < heap[index - 1]:
        heap[parentIndex - 1], heap[index - 1] = heap[index - 1], heap[parentIndex - 1]
        sift_up(heap, parentIndex)

File: sprint5/By lessons/lesson10code1/lesson10code1.java
import java.util.ArrayList;

public class HeapFunctions {
    public static void heapAdd(ArrayList<Integer> heap, int key) {
        int index = heap.size() + 1;
        heap.add(index - 1, key);
        siftUp(heap, index);
    }

    public static void siftUp(ArrayList<Integer> heap, int index) {
        if (index == 1) return;

        int parentIndex = index / 2;
        if (heap.get(parentIndex - 1) < heap.get(index - 1)) {
            int temp = heap.get(parentIndex - 1);
            heap.set(parentIndex - 1, heap.get(index - 1));
            heap.set(index - 1, temp);
            siftUp(heap, parentIndex);
        }
    }

    public static void main(String[] args) {
        ArrayList<Integer> heap = new ArrayList<>();
        heapAdd(heap, 5);
        heapAdd(heap, 8);
        heapAdd(heap, 3);
        System.out.println(heap); // Output: [8, 5, 3]
    }
}
